### ISSUE TITLE
Roslyn incorrectly reports a warning for returning a non-nullable referene to a readonly nullable reference

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#46971]
**Version Used**: 
```
Branch master (18 Aug 2020)
Latest commit ca88e94 by Julien Couvreur:
Handle deconstruction variables in invalid scenarios (#46655)
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  ref readonly string? M1(ref string y)
  {
    return ref y;
  }
}
```

**Expected Behavior**:
No warnings.

**Actual Behavior**:
`CS8619: Nullability of reference types in value of type 'string' doesn't match target type 'string?'.`
Is reported for `M1`

**Notes**
There's nothing wrong with `M1` as the returned reference cannot be used to write a nullable value to a non-nullable location so it's completely safe to return it.
Roslyn even allows you to do so without warnings with an intermediate variable
```
  ref readonly string? M2(ref string y)
  {
    ref readonly var z = ref y;
    return ref z;
  }
```

**Comment by RikkiGibson** [comment]
I do not think we have precedent for allowing covariance of readonly refs. At a glance this feels like something we could allow.

**Comment by TessenR** [comment]
@RikkiGibson 

Allowed in inheritance
```
#nullable enable
class Base
{
  public virtual ref readonly string? M() => throw null!;
}

class Derived : Base
{
  public override ref readonly string M() => throw null!; // no warnings, 'ref' would warn
}
```

Allowed in method group conversions
```
#nullable enable
delegate ref readonly T D<T>();

class C
{
    void M()
    {
        D<string?> d1 = MethodGroup; // no warnings, 'ref' would warn
    }
    
    string f = "";
    ref readonly string MethodGroup() { return ref f; }
}
```

Although it's admittedly supported rather inconsistently e.g. I'm pretty sure this code snippet shouldn't have a warning as well

```
#nullable enable
delegate ref readonly T D<T>();

class C
{
    void M()
    {
        D<string> d1 = MethodGroup;
        D<string?> d2 = d1; // false warning
    }
    
    string f = "";
    ref readonly string MethodGroup() { return ref f; }
}
```

Should I create a separate issue for this one?

**Comment by TessenR** [comment]
I think this one also shouldn't warn but does currently
```
class C
{
    void M(string s)
    {
        ref string r1 = ref s;
        ref readonly string? r2 = ref r1; // here
    }
}
```

**Comment by RikkiGibson** [comment]
Welp, covariance of readonly refs is new to me (is it specific to nullability?). Particularly if we allow it for overrides already, I think we might as well just allow it for conversions in general, since that's all we're really checking.. e.g. is the base method's parameter convertible to the derived method's parameter, and is the derived method's return convertible to the base method's return.

> ```
> #nullable enable
> delegate ref readonly T D<T>();
> 
> class C
> {
>     void M()
>     {
>         D<string> d1 = MethodGroup;
>         D<string?> d2 = d1; // false warning
>     }
>     
>     string f = "";
>     ref readonly string MethodGroup() { return ref f; }
> }
> ```
> 
> Should I create a separate issue for this one?

I suspect this will be solved by #46953 by simply using the same variance checks that overrides use when examining delegate conversions. It would be handy to have a standalone issue for it that we can tag separately.

**Comment by TessenR** [comment]
AFAIK it's specific to nullability. I've created https://github.com/dotnet/roslyn/issues/46977 for the delegate conversion example
